### PR TITLE
Add iree_ref for gemm functionality test

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds utils to generate an iree reference
mmt kernel that can be compiled and run to
obtain a golden output for comparison. Compared
to the torch output, the advantage of this output
is that it can be matched exactly and makes for
an easier comparison target.

### CODE REVIEW

**Comment by harsh-nod** [comment]
This test will fail until https://github.com/iree-org/iree-turbine/pull/101 lands.

File: shark_turbine/kernel/wave/iree_utils.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import torch
+from .utils import compile_and_invoke
+
+
+def get_mmt_asm(lhs_type: str, rhs_type: str, acc_type: str) -> str:
+    acc_dtype = acc_type.split("x")[-1]
```
Comment (raikonenfnu) [inline shark_turbine/kernel/wave/iree_utils.py:6]: [No action for now]: I think in the future we can also update this to use jinja templates S.T we can create separate files for MLIR templates.

File: shark_turbine/kernel/wave/iree_utils.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    return matmul_function
+
+
+def dtype_str(dtype: torch.dtype) -> str:
+    if dtype == torch.float32:
+        return "f32"
+    elif dtype == torch.float16:
+        return "f16"
+    else:
+        raise ValueError(f"Unsupported dtype: {dtype}")
```
Comment (raikonenfnu) [inline shark_turbine/kernel/wave/iree_utils.py:26]: NIT: Can we use https://github.com/iree-org/iree-turbine/blob/36c99a686d5790add276bae3a88aa33dfb23c939/shark_turbine/support/conversions.py#L66-L86 instead?

File: shark_turbine/kernel/wave/iree_utils.py (hunk 1 above)
Comment (Hardcode84) [reply shark_turbine/kernel/wave/iree_utils.py:6]: I think we can already do
```python
matmul_function = f"""
func.func @mmt(%lhs: tensor<{lhs_type}>, %rhs: tensor<{rhs_type}>) 
...
"""

File: shark_turbine/kernel/wave/utils.py [hunk 3]
```diff
@@ -181,3 +190,118 @@ def remove_global_indexing(
...
+    func_name: str,
+    config: dict[str, str],
+    kernel_inputs: list[torch.Tensor],
+    kernel_outputs: list[torch.Tensor],
+    run: bool = False,
+    run_bench: bool = False,
+):
+    backend = config["backend"]
+    device = config["device"]
+    flags = [
```
Comment (raikonenfnu) [inline shark_turbine/kernel/wave/utils.py:243]: Just curious do we not need to specify a different flag when we start compilation from distributed/lower level  vector dialect VS linalg? :O

File: shark_turbine/kernel/wave/iree_utils.py (hunk 1 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/iree_utils.py:6]: Good point on the jinja templates. Thanks for the pointer of the f-triple string, will change to use that.

File: shark_turbine/kernel/wave/iree_utils.py (hunk 2 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/iree_utils.py:26]: sure, good find!

File: shark_turbine/kernel/wave/iree_utils.py (hunk 1 above)
Comment (raikonenfnu) [reply shark_turbine/kernel/wave/iree_utils.py:6]: Ooo, if that is possible, that is also very nice haha

**Comment by raikonenfnu** [comment]
Overall looks awesome! just some NITs and Qs then we should be good to go! :)

File: shark_turbine/kernel/wave/utils.py (hunk 3 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/utils.py:243]: Surprisingly not :) You can call it good design on the iree side :)

Comment (raikonenfnu) [reply shark_turbine/kernel/wave/utils.py:243]: Nice! :D

File: shark_turbine/kernel/wave/iree_utils.py (hunk 1 above)
Comment (Hardcode84) [reply shark_turbine/kernel/wave/iree_utils.py:7]: + you don't need `\n` in multiline literals

**Review by raikonenfnu — APPROVED** [review]
Looks great! feel free to land after Ivan's comment on `\n` for mult-line literal! :)

File: shark_turbine/kernel/wave/iree_utils.py (hunk 1 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/iree_utils.py:7]: thanks you learn something new everyday :)
